### ISSUE TITLE
learn_ml_dl

### ISSUE THREAD

**Issue by NicholasTao** [NicholasTao/NicholasTao.github.io#14]
机器学习简介  |  Machine Learning  |  Google for Developers
https://developers.google.com/machine-learning/crash-course/linear-regression?hl=zh-cn

**Comment by NicholasTao** [comment]
Pytorch内存简析及大模型训练的内存优化方式总结
https://jx.huawei.com/community/comgroup/postsDetails?postId=4184bac71f49460ab9a7cb8a2bbb57bd&noTop=true&type=freePost&welink_open_uri=aDU6Ly80NzE2NTE3MzE0Nzc5NTcvaHRtbC9pbmRleC5odG1sIy9qeC9kZXRhaWw%2FaWQ9NDE4NGJhYzcxZjQ5NDYwYWI5YTdjYjhhMmJiYjU3YmQmdHlwZT1mcmVlX3Bvc3QmdXJsPQ%3D%3D

**Comment by NicholasTao** [comment]
理解语言的 Transformer 模型https://tensorflow.google.cn/tutorials/text/transformer?hl=zh-cn

**Comment by NicholasTao** [comment]
华为mindspore
https://www.mindspore.cn/tutorials/experts/zh-CN/r1.8/parallel/communicate_ops.html#alltoall

**Comment by NicholasTao** [comment]
并行技术
https://colossalai.org/zh-Hans/docs/concepts/paradigms_of_parallelism/

**Comment by NicholasTao** [comment]
一、神经网络训练
MNIST数据集
MLP
正向传播
梯度下降与反向传播
批处理与矩阵运算
参考资料：

知乎：
https://zhuanlan.zhihu.com/p/21747416588
CMU 10714公开课第一节：
https://colab.research.google.com/github/dlsyscourse/hw0/blob/main/hw0.ipynb
(TODO: 国内视频链接）
https://gitee.com/lrwei0709/10714-hw/tree/hw0/
二、词嵌入、Transformer、GPT和DeepSeek
了解：

Word2Vec
[Efficient Estimation of Word Representations in Vector Space](https://arxiv.org/pdf/1301.3781)
Transformer
[Attention Is All You Need](https://arxiv.org/pdf/1706.03762)
GPT
[Improving Language Understanding by Generative Pre-Training](https://cdn.openai.com/research-covers/language-unsupervised/language_understanding_paper.pdf)
DeepSeek-V2
[DeepSeek-V2: A Strong, Economical, and Efficient Mixture-of-Experts Language Model](https://arxiv.org/pdf/2405.04434)
问题：
Transformer的结构是什么，大概有多少参数？
DeepSeek-V2的注意力层有何特别，为什么要这样设计？
DeepSeek-V2的FFN层有何特别？
三、通信原语与并行切分策略
通信原语
Send+Recv / AllReduce / AllGather / ReduceScatter / AllToAll

DP (Data Parallelism)：
[ZeRO: Memory Optimizations Toward Training Trillion Parameter Models](https://dl.acm.org/doi/pdf/10.5555/3433701.3433727)

问题：
什么是global batch size？
ZeRO和普通的DP相比，引入了哪些额外的通信？
ZeRO和普通的DP相比，减少了哪些内存？
TP (Tensor Parallelism)：

基于AllReduce：
[Megatron-LM: Training Multi-Billion Parameter Language Models Using Model Parallelism](https://arxiv.org/pdf/1909.08053)
基于AllGather + ReduceScatter：
[Reducing Activation Recomputation in Large Transformer Models](https://arxiv.org/pdf/2205.05198)
问题：
TP引入了哪些通信，通信量是多少？
TP对模型的静态内存有什么影响？
TP对模型的动态内存有什么影响？
PP (Pipeline Parallelism)：

最原始的Pipeline排布
[GPipe: Easy Scaling with Micro-Batch Pipeline Parallelism](https://arxiv.org/pdf/1811.06965)
对GPipe的1F1B改进：
[Memory-Efficient Pipeline-Parallel DNN Training](https://arxiv.org/pdf/2006.09503)
1F1B + VPP改进：
[Efficient Large-Scale Language Model Training on GPU Clusters Using Megatron-LM](https://arxiv.org/pdf/2104.04473)
1F1B + ZBB改进：
[ZERO BUBBLE PIPELINE PARALLELISM](https://arxiv.org/pdf/2401.10241)
问题：
什么是micro batch size？
1F1B相较于GPipe的主要改进是什么？
VPP相较于1F1B的主要改进是什么？
ZBB相较于1F1B的主要改进是什么？
PP对模型的静态内存有什么影响？
PP对模型的动态内存有什么影响？
PP的空泡率怎么计算？
MoE / EP (Expert Parallelism)：
[A Hybrid Tensor-Expert-Data Parallelism Approach to Optimize Mixture-of-Experts Training](https://arxiv.org/pdf/2303.06318)
[MoE Parallel Folding: Heterogeneous Parallelism Mappings for Efficient Large-Scale MoE Model Training with Megatron Core](https://arxiv.org/html/2504.14960v1)

MindSpeed All2All 版本MoE：
https://gitee.com/ascend/MindSpeed/blob/master/mindspeed/core/transformer/moe/moe_layer_overlap_all2all.py
MindSpeed AllGather 版本MoE：
https://gitee.com/ascend/MindSpeed/blob/master/mindspeed/core/transformer/moe/moe_layer_overlap_allgather.py
MindSpeed MC2 版本MoE：
https://gitee.com/lrwei0709/MindSpeed/blob/br_gmm_a2a_more/mindspeed/core/transformer/moe/moe_layer_fused_gmm_all2all.py
DualPipe / DualPipeV：
[DeepSeek-V3 Technical Report](https://arxiv.org/html/2412.19437v1)
[双流并行(DualPipe) 没有双流会更好](https://hackmd.io/@ufotalent/S1N_ay0ckx)
问题：
EP引入了哪些通信？
EP对模型的静态内存有什么影响？
EP对模型的动态内存有什么影响？
DualPipe有气泡，为什么不使用无气泡的ZBB？
CP (Context Parallelism)：
[Online normalizer calculation for softmax](https://arxiv.org/pdf/1805.02867)
[Ring Attention with Blockwise Transformers for Near-Infinite Context](https://arxiv.org/pdf/2310.01889)

四、实现
hiascend网站
aclnn
torch_npu
MindSpeed-LLM

**Comment by NicholasTao** [comment]
更简单的方式来解释反向传播！可以把它想象成一个 学习纠错过程，类似于学生做题并通过老师的反馈来提高成绩。

假设场景：
你在练习投篮球，目标是投进篮筐。但第一次投的时候，球偏左了，没进。这时，你需要调整你的投篮姿势，让下次投得更准确。

反向传播的步骤：
试一试（前向传播）：

你先随便投篮（神经网络做预测）。

发现投歪了，没进（计算误差，即损失函数）。

找到问题（计算梯度）：

你观察自己的动作，发现是手腕没用力，导致球飞行方向不对（反向传播计算误差的影响）。

调整动作（参数更新）：

你稍微改变手腕力度，下次投篮就会更准一些（权重更新）。

不断练习（重复训练）：

你不断练习投篮，每次调整，直到命中率提高（神经网络经过多次更新，最终学会正确的模式）。

反向传播的本质就是一种 试错学习，让神经网络不断调整自己，直到可以正确完成任务。

**Comment by NicholasTao** [comment]
https://zhuanlan.zhihu.com/p/30201040247
全景解读 LLM 后训练技术

**Comment by NicholasTao** [comment]
DeepSeek-R1 工作原理
让模型进行慢思考
思维链
（Chain of Thought）
在不损失能力的情况下缩小模型
蒸馏
（Distillation）
让模型自我探索和训练
强化学习
（Reinforcement Learning）
⚫ DeepSeek R1论文：https://github.com/deepseek-ai/DeepSeek-R1/blob/main/DeepSeek_R1.pdf
⚫ DeepSeek R1论文图解：https://zhuanlan.zhihu.com/p/20844750193

**Comment by NicholasTao** [comment]
https://redresscompliance.com/10-key-differences-between-deepseek-and-other-ai-models/
DeepSeek 与其他 AI 模型的 10 个主要区别
https://www.yicaiai.com/news/article/67a6d92c4ddd79f11a000e92
深入探索Deepseek：多跳推理能力优化之道
